test/chunking-form/samples/dynamic-import-treeshaking/_expected/es/dep2.js

```diff
@@ -0,0 +1,7 @@
+import { multiplier } from './chunk-36864708.js';
+
+function mult (num) {
+  return num + multiplier;
+}
+
+export { mult };
```

Comment:
So I assume the point of this test is to show that the unused imports from `dep2.js` are properly removed. But maybe we can do better by not emitting this chunk at all? But this might also be a possible future enhancement.

For instance the logic could be that a chunk is not emitted when
* There are no synchronous imports from it (otherwise we should probably include it for potential side-effects until we found away to improve the side-effect detection logic to find out if a chunk has side-effects in its initialization code. Actually, this could probably be done by checking if any of the top-level statements are included or if one of the imports has side-effects. But again, maybe a future enhancement)
* None of the dynamic imports are included

Another interesting fact I stumbled upon that we might improve in the future is that when
* a chunk exports a variable and
* this chunk is imported by two other chunks both of which import this variable and
* only one of the chunks actually uses this variable,
the imported variable is retained for both chunks. The reason is that there is only one variable entity and as soon as it is included, all imports of this variable are included. If none of the chunks use the variable, it is properly removed. I think it is more of a cosmetic issue but still.

Reply:
You've exactly understood the two limitations of the approach I put together - the chunking algorithm is run on the graph _before_ treeshaking, and the export boundaries between chunks aren't properly pruned. I originally aimed to handle both of these cases (and my first few attempts were in this direction), but hit some road blocks.

When running chunking on the graph after treeshaking, execution order still has to be maintained as an invariant. In addition we need to work out which chunks can be considered pure in order to skip emission like you mention.

If I remember correctly, the main issue I hit when implementing this (after fixing up export tracing etc etc) was that the concept of "marked" or "used" must then be considered uniquely for each specific chunk, which wasn't a simple capability to add. I can't remember what the other issues were, but if you're interested I could try trawling through my notes for future reference.

Reply:
Note that in may of my practical investigations of the algorithm, these assumptions ended up not seeming so bad at all though. The majority of cases work really well without these features. I think as we use this feature we'll get a better idea for how bad "pure" chunk relations and unpruned boundaries might end up looking, and can either consider specific fixes, or taking the more general route then.